torch/_functorch/_aot_autograd/subclass_utils.py

```diff
@@ -111,29 +126,82 @@ def create_subclass_meta(
 # Instead, we specialize the logic for the inference vs. joint graph case.
 # NOTE: this function is hot, since we unwrap tensor subclass inputs at runtime
-def unwrap_tensor_subclasses(wrapped_args, *, is_joint_structure: bool):
-    def concat_inner_tensors_from_subclasses(xs):
-        xs_inner = []
-        for x in xs:
-            if is_traceable_wrapper_subclass(x):
-                xs_inner.extend(get_plain_tensors(typing.cast(Tensor, x)))
-            else:
-                xs_inner.append(x)
-        return xs_inner
 
-    if is_joint_structure:
-        assert isinstance(wrapped_args, tuple) and len(wrapped_args) == 2
-        assert isinstance(wrapped_args[0], (tuple, list)) and isinstance(
-            wrapped_args[1], (tuple, list)
-        )
-        unwrapped_args_fw = concat_inner_tensors_from_subclasses(wrapped_args[0])
-        unwrapped_args_tangents = concat_inner_tensors_from_subclasses(wrapped_args[1])
-        unwrapped_args = (unwrapped_args_fw, unwrapped_args_tangents)
-    else:
-        assert isinstance(wrapped_args, (list, tuple))
-        unwrapped_args_fw = concat_inner_tensors_from_subclasses(wrapped_args)
-        unwrapped_args = unwrapped_args_fw
-    return unwrapped_args
+
+def filter_symints(lst: Iterable[Union[int, SymInt]]):
+    # Capture all SymInts from the iterable. A SymInt is not included if it is
+    # a nested int and filter_nested_int is True
+    def symint_check(s: Union[int, SymInt]) -> bool:
+        return isinstance(s, SymInt) and not s.node.is_nested_int()
+
+    return [s for s in lst if symint_check(s)]
+
+
+def compute_symint_placeholders(lst: Iterable[Union[None, int, SymInt]]) -> List[bool]:
+    # Non-nested symints are replaced with None in `make_runtime_safe()`
+    return [s is None for s in lst]
+
+
+# The reason for "append_symints"
+#
+# * At compile time: we append extra symint args when unwrapping primals
+# (but not tangents, because they should always share symints with primals).
+# We also append extra symints when unwrapping the subclass outputs of the
+# traced function, so we can return them as extra outputs
+#
+# * At runtime: we similarly append subclass sizes when we unwrap subclass
+# primals (but not tangents) on entry to the forward.
+#
+# subclass_metas is needed at runtime to compute which indices are symints in
+# the outer_size/outer_stride
+def unwrap_tensor_subclasses(
+    wrapped_args: List[Union[Tensor, int]],
+    *,
+    is_runtime: bool,
+    append_symints: bool,
+    subclass_metas: Optional[List[Union[int, SubclassCreationMeta]]] = None,
+):
+    xs_inner: List[Union[int, Tensor, SymInt]] = []
+
+    for idx, x in enumerate(wrapped_args):
+        if not is_traceable_wrapper_subclass(x):
+            xs_inner.append(x)
+            continue
+
+        size = x.size()
+        stride = x.stride()
+        xs_inner.extend(get_plain_tensors(typing.cast(Tensor, x)))
+        if append_symints:
+            if is_runtime:
```

Comment:
FYI @IvanKobzarev - subclass runtime overhead has come up (see https://discord.com/channels/1189498204333543425/1293570158425538602/1293941322104832011) - and this PR is probably going to make it a bit worse.

I think I'd rather just land this PR and get dynamic shapes working first, then worry too much about the overhead directly in this PR. But what I imagine is that right now we have a generic function that takes in an `is_runtime` flag to branch on - and we're probably going to need to split it off into a trace-time vs. runtime-only function for doing the unwrapping, that is carefully written to not be slow.